### PR TITLE
The Encoding Update

### PR DESCRIPTION
**Pull request by Minty-Meeo**
While Nintendo's own documents claim GameCube and Wii disc file symbol tables only support 7-bit ASCII, this is far from the truth.  Indeed, even some first-party Nintendo games shipped with Shift-JIS encoded file symbol tables.  My guess?  The locale of whatever Windows machine mastered a GameCube or Wii disc influenced how wide character strings (UCS-2) were converted to narrow character strings.  To account for all possibilites, this update adds extensible multi-byte character set options to NOD-Tool.

A rundown of notable changes:
 - "-c XXXXX" option added to set the encoding of the GameCube / Wii ISO(s) being processed.
 - "SystemStringConv" renamed to "DiscLocToSystemConv"
 - "SystemUTF8Conv" renamed to "SystemToDiscLocConv"
 - Help message updated with new info.
 - Bugfix: AddBuildName had a logic error wherein the length of the SystemString was being used instead of length of the disc locale string.  This would corrupt the File Symbol Table if the disc locale string's length was greater than the SystemString's length.
 - Bugfix: recursiveMergeFST was not keeping track of parent indexes at all, meaning nested folders and their contents would be corrupted.  I simply copied the way recursiveBuildFST did things to fix this.
 - Bugfix (Windows): On Windows, for some reason, Sstat was a typedef for _stat (32-bit) instead of _stat64 (64-bit).  This is confounding, because untrimmed Wii ISOs will always be larger than the unsigned 32-bit integer limit (4,699,979,776 bytes vs 4,294,967,295 bytes), meaning the MergeWii errand has never worked for untrimmed ISOs on Windows.  Was this never tested??
 - Bugfix (Windows): Did you know Windows Command Prompt fully supports Unicode?  Stdio streams are now in _O_U16TEXT mode for Windows only.  Previously, attempting to print any character that could not be narrowed to your locale's encoding would either silently fail (std functions), or throw an exception (fmt functions).  As a minor drawback, narrow character print functions can no longer be used when stdio is in _O_U16TEXT mode, necessitating my PR for Logvisor here: (https://github.com/AxioDL/logvisor/pull/7)
 - ExtractionContext::progressCB now uses SystemStringView because widechar printing works correctly on Windows now.
 - progFunc lambda no longer throws exceptions when printing unicode because widechar printing works correctly on Windows now.
 - Top-level constructors and functions with a Codepage_t parameter have also signatures that default to the US-ASCII codepage.
    - DiscGCN constructor
    - DiscBuilderGCN constructor
    - DiscBuilderGCN::CalculateTotalSizeRequired
    - DiscMergerGCN constructor
    - DiscMergerGCN::CalculateTotalSizeRequired
    - DiscWii constructor
    - DiscBuilderWii constructor
    - DiscBuilderWii::CalculateTotalSizeRequired
    - DiscMergerWii constructor
    - DiscMergerWii::CalculateTotalSizeRequired
    - OpenDiscFromImage
 - Conversion between system encoding and disc locale encoding has checks in place to warn the user if string conversion goes awry.

### CODE REVIEW

**Comment by Minty-Meeo** [comment]
As mentioned, this PR does not work without the changes I also made to Logvisor, so please check that PR first.

**Comment by Antidote** [comment]
While I'm not a huge fan of the use of iconv, this PR is a much better approach than your original. I'm going to go ahead and approve this but I'd like @lioncash to do a code review before it gets merged when you think it's ready.

**Comment by Minty-Meeo** [comment]
The inclusion of iconv has only increased the size of the Linux release executable by 5.56% according to my test (712,656 bytes w/ iconv vs. 675,120 bytes w/o iconv).  Hopefully this relieves your concern about bloat?

**Comment by Antidote** [comment]
The bloat I was concerned about wasn't binary size, but porting surface, I'm perfectly fine porting my own code to various platforms as needed, but I'm not really interested in spelunking through iconv's rather sizeable codebase to fix issues if that was to happen. That being said iconv has a rather large userbase so the probability of it getting ported to a new platform is rather high.

**Comment by Minty-Meeo** [comment]
I've never worked with submodules before.  How do I now update the Logvisor submodule in this repository to be up-to-date with the changes I made in the Logvisor repository?

**Comment by Antidote** [comment]
You can do it with this series of commands:
```
git submodule sync logvisor
git add logvisor
```
Then committing as usual.

File: driver/main.cpp
Comment (Antidote) [inline driver/main.cpp]: Any way we can unify this instead of having two different ways?

File: driver/main.cpp
Comment (Antidote) [inline driver/main.cpp]: prefix increment

File: driver/main.cpp
Comment (Antidote) [inline driver/main.cpp]: Same here

File: driver/main.cpp
Comment (Antidote) [inline driver/main.cpp]: Looks like you use a lot of postfix increment, please normalize to prefix increment for non-assignments.

File: driver/main.cpp
Comment (Antidote) [inline driver/main.cpp]: Use camelCase rather than snake_case for variables.

File: include/nod/Util.hpp
Comment (Antidote) [inline include/nod/Util.hpp:79]: Go ahead and change these declarations to `using name = type`

File: include/nod/Util.hpp
Comment (Antidote) [inline include/nod/Util.hpp:99]: Same here

**Review by Antidote — CHANGES_REQUESTED** [review]
Please make the following changes

File: driver/main.cpp
Comment (Minty-Meeo) [reply driver/main.cpp]: http://site.icu-project.org/
Would it be worth using this library?  I think it can replace Windows stringapi and Linux's iconv.

Comment (Antidote) [reply driver/main.cpp]: It uses a configure script rather than CMake, if you think you can get it cleanly integrated, I like the API better, and it's more sensibly organized.

File: driver/main.cpp
Comment (lioncash) [inline driver/main.cpp]: This should be joined with the previous line. Ditto elsewhere.

File: driver/main.cpp
Comment (lioncash) [inline driver/main.cpp]:
```suggestion
  if (errand.empty()) {
```

File: driver/main.cpp
Comment (lioncash) [inline driver/main.cpp]:
```suggestion
    while (argidx < argc) {
```

File: driver/main.cpp
Comment (lioncash) [inline driver/main.cpp]:
```suggestion
    while (argidx <argc ) {
```

File: driver/main.cpp
Comment (lioncash) [inline driver/main.cpp]:
```suggestion
    while (argidx < argc) {
```

File: driver/main.cpp
Comment (lioncash) [inline driver/main.cpp]:
```suggestion
    while (argidx < argc) {
```

**Comment by Antidote** [comment]
Any updates on this?

**Comment by Minty-Meeo** [comment]
I found a CMake wrapper for the International Components for Unicode library, but lost interest as I do not know anything about CMake to even begin to figure that out.  Would it be acceptable if I cut down the codepage selection to whatever IDs WinAPI can support, then make a std::map for the equivalent iconv codepage strings?

**Comment by Minty-Meeo** [comment]
"using Sstat = struct _stat64" appears to do the same thing as "typedef struct _stat64 Sstat" according to Visual Studio's Intellisense, but trying to compile it produces a mountain of errors.  I am going to leave the typedefs alone for now.

**Comment by Minty-Meeo** [comment]
I am going to take another stab at the wrapper I mentioned: [https://github.com/viaduck/icu-cmake](https://github.com/viaduck/icu-cmake).  I think I got it set up in the CMakeLists right this time.

**Comment by Minty-Meeo** [comment]
I can't figure this out.  How important is unifying the codepage selection across Windows and Linux?  It turns out making a std::map isn't so simple either, given gaps for matching codepages, as well as multiple potential aliases that would need testing.

**Comment by Antidote** [comment]
We only care about the exposed ones, don't worry about mapping everything, just the ones we're immediately concerned with.

**Comment by Antidote** [comment]
> "using Sstat = struct _stat64" appears to do the same thing as "typedef struct _stat64 Sstat" according to Visual Studio's Intellisense, but trying to compile it produces a mountain of errors. I am going to leave the typedefs alone for now.

The reason the alias isn't working is because you have to use `struct`

**Comment by Minty-Meeo** [comment]
Where do I put "struct".  Nothing works.

``using Sstat = _stat64;``
function "_stat64" is not a type name

``struct using Sstat = _stat64;``
expected either a definition or a tag name

``using struct Sstat = _stat64;``
expected an identifier

``using Sstat struct = _stat64;``
expected a ';'
identifier "Sstat" is undefined

``using Sstat = struct _stat64;``
> trying to compile it produces a mountain of errors.

``using Sstat = _stat64 struct;``
expected either a definition or a tag name
function _stat64 is not a type name
invalid combination of type specifiers

**Comment by Antidote** [comment]
you missed my point, you *can't* use a `using` alias *because* you need the `struct` type specifier. This is due to _stat64 not being fully declared in the header. Just stick with the typedef.

That being said you could try `using Sstat = typename struct _stat64;` but I doubt that'll work

**Comment by Minty-Meeo** [comment]
I hope I didn't messed up the PR too much while trying to fix Logvisor in my fork.

**Comment by Minty-Meeo** [comment]
Should I just ``git reset master`` and then force push this all as one commit?

**Comment by Antidote** [comment]
Go ahead and squash this into one commit and I'll merge tomorrow after testing.

**Review by Antidote — APPROVED** [review]

**Comment by Antidote** [comment]
everything looks good! Thanks for giving this another shot.

**Pull request merged**
